pkg/controller/persistentpodstate/persistent_pod_state_controller.go

```diff
@@ -0,0 +1,451 @@
+	for i := range pods {
+		pod := pods[i]
+		if pod.DeletionTimestamp.IsZero() {
+			matchedPods[pod.Name] = pod
+		}
+	}
+	return matchedPods, inner, nil
+}
+
+func (r *ReconcilePersistentPodState) createPodState(pod *corev1.Pod, nodeTopologyKeys sets.String) (appsv1alpha1.PodState, error) {
+	// pod state
+	podState := appsv1alpha1.PodState{
+		NodeTopologyLabels: map[string]string{},
+	}
+	//get node of pod
+	node := &corev1.Node{}
+	err := r.Get(context.TODO(), client.ObjectKey{Name: pod.Spec.NodeName}, node)
+	if err != nil {
+		klog.Errorf("fetch pod(%s/%s) node(%s) error %s", pod.Namespace, pod.Name, pod.Spec.NodeName, err.Error())
+		return podState, err
+	}
+	podState.NodeName = pod.Spec.NodeName
+	for _, key := range nodeTopologyKeys.List() {
+		podState.NodeTopologyLabels[key] = node.Labels[key]
+	}
+	return podState, nil
+}
+
+func isInStatefulSetReplicas(index int, sts *innerStatefulset) bool {
+	reserveOrdinals := sets.NewInt(sts.ReserveOrdinals...)
+	replicas := sets.NewInt()
+	replicaIndex := 0
+	for realReplicaCount := 0; realReplicaCount < int(sts.Replicas); replicaIndex++ {
+		if reserveOrdinals.Has(replicaIndex) {
+			continue
+		}
+		realReplicaCount++
+		replicas.Insert(replicaIndex)
+	}
+	return replicas.Has(index)
+}
+
+// auto generate PersistentPodState crd
+func (r *ReconcilePersistentPodState) autoGeneratePersistentPodState(req ctrl.Request) error {
+	// req.Name Format = generate#{apiVersion}#{workload.Kind}#{workload.Name}
+	// example for generate#apps/v1#StatefulSet#echoserver
+	arr := strings.Split(req.Name, "#")
+	if len(arr) != 4 {
+		klog.Warningf("Reconcile PersistentPodState workload(%s) is invalid", req.Name)
+		return nil
+	}
+	// fetch workload
+	apiVersion, kind, ns, name := arr[1], arr[2], req.Namespace, arr[3]
+	workload, err := r.finder.GetScaleAndSelectorForRef(apiVersion, kind, ns, name, "")
+	if err != nil {
+		return err
+	} else if workload == nil {
+		klog.Warningf("Reconcile PersistentPodState workload(%s) is Not Found", req.Name)
+		return nil
+	}
+	// fetch persistentPodState crd
+	oldObj := &appsv1alpha1.PersistentPodState{}
+	err = r.Client.Get(context.TODO(), client.ObjectKey{Namespace: ns, Name: name}, oldObj)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		// not found
+		oldObj = nil
+	}
+
+	// auto generate persistentPodState crd object
+	if workload.Metadata.Annotations[appsv1alpha1.AnnotationAutoGeneratePersistentPodState] == "true" {
+		if workload.Metadata.Annotations[appsv1alpha1.AnnotationRequiredPersistentTopology] == "" &&
+			workload.Metadata.Annotations[appsv1alpha1.AnnotationPreferredPersistentTopology] == "" {
+			klog.Warningf("statefulSet(%s/%s) persistentPodState annotation is incomplete", workload.Metadata.Namespace, workload.Name)
+			return nil
+		}
+
+		newObj := newStatefulSetPersistentPodState(workload)
+		// create new obj
+		if oldObj == nil {
+			if err = r.Create(context.TODO(), newObj); err != nil && errors.IsAlreadyExists(err) {
```

Comment:
Does it return err when it is AlreadyExists?

```golang
    if err = r.Create(context.TODO(), newObj); err != nil
        if errors.IsAlreadyExists(err) {
            return nil
        }
        return err
    }
```

to avoid logging "create StatefulSet" when it gets AlreadyExists error.